[jit] Convenience APIs for script objects

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20226 [jit] Convenience APIs for script objects**

The compiler should be emitting static lookups to object slots, but
people using the C++ API should have some sugar for doing attribute
lookup.

Differential Revision: [D15243625](https://our.internmc.facebook.com/intern/diff/D15243625)

suo committed:
[jit] Convenience APIs for script objects

The compiler should be emitting static lookups to object slots, but
people using the C++ API should have some sugar for doing attribute
lookup.

suo committed:
Update on "[jit] Convenience APIs for script objects"

[jit] Convenience APIs for script objects

The compiler should be emitting static lookups to object slots, but
people using the C++ API should have some sugar for doing attribute
lookup.

gh-metadata: pytorch pytorch 20226 gh/suo/36/head

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -113,6 +113,16 @@ std::string ivalue::Object::name() const {
   return this->type_->qualname();
 }
 
+IValue ivalue::Object::getAttr(const std::string& name) const {
```
Probably want the same api as pybind11, which uses `attr`

suo replied:
pybind's attr returns a functor which lets you write to the object by assigning to it. I think that's probably a little overkill in this case.

I think it'll expose static `getattr` and `setattr` like the python builtins

TheCodez replied:
The old API seems simpler, this makes it more like a C API. Maybe we shouldn't try to match pythons semantics here?

suo replied:
I'm okay with that. whatever is more natural for people I guess

zdevito commented on test/cpp/jit/test_class_import.h:
```diff
@@ -58,6 +63,32 @@ void testClassImport() {
+      constantTable,
+      nullptr);
+
+  // Incorrect arguments for constructor should throow
```
throooooow

suo replied:
damn the macbook keyboard :(

suo committed:
Update on "[jit] Convenience APIs for script objects"

[jit] Convenience APIs for script objects

The compiler should be emitting static lookups to object slots, but
people using the C++ API should have some sugar for doing attribute
lookup.

gh-metadata: pytorch pytorch 20226 gh/suo/36/head

suo committed:
Update on "[jit] Convenience APIs for script objects"

[jit] Convenience APIs for script objects

The compiler should be emitting static lookups to object slots, but
people using the C++ API should have some sugar for doing attribute
lookup.

gh-metadata: pytorch pytorch 20226 gh/suo/36/head

suo committed:
Update on "[jit] Convenience APIs for script objects"

[jit] Convenience APIs for script objects

The compiler should be emitting static lookups to object slots, but
people using the C++ API should have some sugar for doing attribute
lookup.

gh-metadata: pytorch pytorch 20226 gh/suo/36/head
